R/discrete_lognormal.R

```diff
@@ -0,0 +1,118 @@
+#' @name discrete_lognormal
+#' @title Discrete lognormal distribution
+#'
+#' @description a discretised lognormal distribution (i.e. sampled by applying 
+#'   the round operation to samples from a lognormal). Due to the numerical 
+#'   instability of integrating across the distribution, a vector of breaks 
+#'   must be defined and the observations will be treated as censored 
+#'   within those breaks
+#'
+#' @param meanlog unconstrained parameters giving the mean of the distribution 
+#'   on the log scale
+#' @param sdlog unconstrained parameters giving the standard deviation of the 
+#'   distribution on the log scale
+#' @param breaks a vector of breaks; observations will be treated as censored 
+#'   within those breaks
+#' @param dim a scalar giving the number of rows in the resulting greta array
+#'
+#' @importFrom R6 R6Class
+#' @export
+ 
+discrete_lognormal <- function(meanlog, sdlog, breaks, dim = NULL) {
+  distrib("discrete_lognormal", meanlog, sdlog, breaks, dim)
+}
+
+# define the discrete lognormal distribution
+discrete_lognormal_distribution <- R6Class(
+  classname = "discrete_lognormal_distribution",
+  inherit = distribution_node,
+  public = list(
+    
+    breaks = NA,
+    lower_bounds = NA,
+    upper_bounds = NA,
+    
+    initialize = function(meanlog, sdlog, breaks, dim) {
+      
+      meanlog <- as.greta_array(meanlog)
+      sdlog   <- as.greta_array(sdlog)
+      
+      # check length of breaks
+      if (length(breaks) <= 1) {
+        msg <- cli::format_error(
+          c(
+            "{.var breaks} must be a vector with at least two break points",
+            "but {.var breaks} has length {length(breaks)}"
+          )
+        ) 
+        stop(
+          msg,
+          call. = FALSE
+        )
+      }
```

Comment:
Can you write this as a function? Perhaps:

```r
check_break_length <- function(breaks){
        if (length(breaks) <= 1) {
        msg <- cli::format_error(
          c(
            "{.var breaks} must be a vector with at least two break points",
            "but {.var breaks} has length {length(breaks)}"
          )
        ) 
        stop(
          msg,
          call. = FALSE
        )
      }
}
```

And then put this in a new file called `checks.R`, and call it as:

```r
check_break_length(breaks)
```

Reply:
Done in [b9a412f](https://github.com/greta-dev/greta.distributions/pull/16/commits/b9a412f24e3fa5bb37ee8a4b063888e68c5efdf0) 😃